### PR TITLE
Add ensure clause check whenever contractInfo is needed

### LINKED ISSUE #17793: Implement recomputation of ensure clause at upgrade time in engine
We currently do not recompute the value of the ensure clause at upgrade time. See https://github.com/digital-asset/daml/pull/17735/files
See: ensureClauseBecomesInvalid, onlyV2EnsureClauseRequiredForImplicitUpgrade and ensureClauseDowngradeToNoLongerValid.

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Resolves #17793

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -860,47 +860,52 @@ private[lf] final class Compiler(
...
-                              t.SCPNone,
-                              let(
-                                env,
-                                translateExp(env.bindExprVar(tmpl.param, tmplArgPos), tmplKey.body),
-                              ) { (keyPos, env) =>
-                                SBSome(translateKeyWithMaintainers(env, keyPos, tmplKey))
-                              },
+      checkPreCondition(env, tmplId, env.toSEVar(tmplArgPos)) { env =>
+        // We use a chain of let bindings to make the evaluation order of SBuildContractInfoStruct's arguments is
+        // independent from the evaluation strategy imposed by the ANF transformation.
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:865]: Maybe a good idea to put the `checkPreCondition`, bellow the comment as is is basically a let binding

**Review by remyhaemmerle-da — APPROVED** [review]
Thanks !

**Comment by samuel-williams-da** [comment]
Given these extra evaluations, is it worth running a quick performance check?

**Review by remyhaemmerle-da — COMMENTED** [review]
The test fixes LGTM.

**Pull request merged**